include/dlaf/common/range2d.hpp

```diff
@@ -0,0 +1,123 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+/// @file
+
+#pragma once
+
+#include "dlaf/common/index2d.h"
+#include "dlaf/types.h"
+
+namespace dlaf {
+namespace common {
+
+namespace internal {
+
+// A base class for column/row-major Iterators.
+class Iterator2DBase {
+public:
+  Iterator2DBase(SizeType ld, SizeType i) : ld(ld), i(i) {}
+  void operator++() noexcept {
+    ++i;
+  }
+
+  bool operator!=(Iterator2DBase const& o) const noexcept {
+    return i != o.i;
+  }
+
+protected:
+  SizeType ld;
+  SizeType i;
+};
+
+}
+
+// An Iterator returning indices in row-major order
+template <typename IndexT, class Tag>
+class RowMajorIterator : public internal::Iterator2DBase {
+  using index2d_t = Index2D<IndexT, Tag>;
+  using Iterator2DBase::i;
+  using Iterator2DBase::ld;
+
+public:
+  RowMajorIterator(SizeType ld, SizeType i) : internal::Iterator2DBase(ld, i) {}
+  index2d_t operator*() const noexcept {
+    return index2d_t(i / ld, i % ld);
```

Comment:
I considered using `dlaf::commo::Ordering` but decided against it because of the performance penalty for a tun-time check. Instead I decided to write to separate iterators because in our code, we usually know at compile time whether we want to iterate in column-major or row major.

I also considered using `computeCoords`, but decided against it as this is simple enough and easy to read as it is and because the function makes a runtime decision based on `dlaf::commo::Ordering`.

Reply:
I can add additional API: iterateRange() that accepts `dlaf::common:Ordering` and calls `iterateRowMajor()` or `iterateColMajor()` respectively.

Reply:
Does it really matter in which order you iterate?
If you use this iterator you shouldn't care.

Reply:
> Does it really matter in which order you iterate?
If you use this iterator you shouldn't care.

The order only matters if the layout of the object that is being accessed requires a particular iteration order for maximum performance. In our case Tiles are stored in column-major and for iterating the grid of tiles of a Matrix it doesn't seem to matter. If we decide that iterating only in Column-Major is sufficient, certain simplifications to the interface can be made. For example, I can make the `Size` type to be Iterable directly, the interface will be simplified.

Reply:
I'm sorry I misunderstood the thing.

I was separating two concepts:
1. how to iterate over a grid;
2. what layout the grid has

But here the memory layout (2) does not come into play, it is just a concept tied to the grid and not to the memory.

Reply:
>I also considered using computeCoords, but decided against it as this is a simple enough implementation that is easy to read and because the function makes a runtime decision based on dlaf::commo::Ordering.

About this, I'm still not sure it is the best choice. I think that here we are in the exactly same problem of `computeCoords`. I'm addressing this in other PRs (see #190 and #183).
I would go for using it here, in case, we may decide to allow the branching at compile time, if the ordering parameter is known as it is in this case.

Reply:
I think we should actually do the reverse. We should refactor `computeCoords` into two separate functions for row-major and column-major order and we should have computeCoords call either depending on `dlaf::common::Ordering`. That way users that know the ordering at compile time don't have to pay any runtime costs.

The functions can be called e.g. `computeRowMajorCoords()` and `computeColMajorCoords()`.

Reply:
Yes, I was not clear enough 😁.

I meant to say that we should use `computeCoords` here, and extend `computeCoords` to be able to branch at compile time with two different implementations, as I think you are saying.